File: sdks/typescript/src/builders.ts
/**
 * Fluent builder for constructing telemetry events
 */

import { v4 as uuidv4 } from 'uuid';
import { TelemetryEvent, PromptInfo, ResponseInfo } from './types';
import { ValidationError } from './errors';

/**
 * Fluent builder for TelemetryEvent
 *
 * @example
 * ```typescript
 * const event = new TelemetryEventBuilder()
 *   .service('my-service')
 *   .model('gpt-4')
 *   .prompt({ text: 'Hello', tokens: 5 })
 *   .response({ text: 'Hi!', tokens: 10, finishReason: 'stop' })
 *   .latency(150.5)
 *   .cost(0.001)
 *   .build();
 * ```
 */
export class TelemetryEventBuilder {
  private _eventId?: string;
  private _timestamp?: string;
  private _serviceName?: string;
  private _traceId?: string;
  private _spanId?: string;
  private _model?: string;
  private _prompt?: PromptInfo;
  private _response?: ResponseInfo;
  private _latencyMs?: number;
  private _costUsd?: number;
  private _metadata: Record<string, string> = {};
  private _errors: string[] = [];

  /**
   * Set event ID (auto-generated if not set)
   */
  eventId(id: string): this {
    this._eventId = id;
    return this;
  }

  /**
   * Set timestamp (auto-generated as current time if not set)
   */
  timestamp(timestamp: string | Date): this {
    if (timestamp instanceof Date) {
      this._timestamp = timestamp.toISOString();
    } else {
      this._timestamp = timestamp;
    }
    return this;
  }

  /**
   * Set service name
   */
  service(name: string): this {
    this._serviceName = name;
    return this;
  }

  /**
   * Set trace ID for distributed tracing
   */
  traceId(id: string): this {
    this._traceId = id;
    return this;
  }

  /**
   * Set span ID
   */
  spanId(id: string): this {
    this._spanId = id;
    return this;
  }

  /**
   * Set model identifier
   */
  model(model: string): this {
    this._model = model;
    return this;
  }

  /**
   * Set prompt information
   */
  prompt(prompt: PromptInfo | { text: string; tokens: number; embedding?: number[] }): this {
    this._prompt = prompt as PromptInfo;
    return this;
  }

  /**
   * Set response information
   */
  response(
    response: ResponseInfo | { text: string; tokens: number; finishReason: string; embedding?: number[] }
  ): this {
    // Map finishReason to finish_reason for convenience
    const responseObj = response as any;
    if (responseObj.finishReason && !responseObj.finish_reason) {
      responseObj.finish_reason = responseObj.finishReason;
      delete responseObj.finishReason;
    }
    this._response = responseObj as ResponseInfo;
    return this;
  }

  /**
   * Set latency in milliseconds
   */
  latency(ms: number): this {
    this._latencyMs = ms;
    return this;
  }

  /**
   * Set cost in USD
   */
  cost(usd: number): this {
    this._costUsd = usd;
    return this;
  }

  /**
   * Add metadata key-value pair
   */
  addMetadata(key: string, value: string): this {
    this._metadata[key] = value;
    return this;
  }

  /**
   * Set metadata object (replaces existing metadata)
   */
  metadata(metadata: Record<string, string>): this {
    this._metadata = { ...metadata };
    return this;
  }

  /**
   * Add an error message
   */
  addError(error: string): this {
    this._errors.push(error);
    return this;
  }

  /**
   * Set errors array (replaces existing errors)
   */
  errors(errors: string[]): this {
    this._errors = [...errors];
    return this;
  }

  /**
   * Build the TelemetryEvent
   * @throws {ValidationError} If required fields are missing
   */
  build(): TelemetryEvent {
    // Validate required fields
    if (!this._serviceName) {
      throw ValidationError.missingField('service_name');
    }

    if (!this._model) {
      throw ValidationError.missingField('model');
    }

    if (!this._prompt) {
      throw ValidationError.missingField('prompt');
    }

    if (!this._response) {
      throw ValidationError.missingField('response');
    }

    if (this._latencyMs === undefined) {
      throw ValidationError.missingField('latency_ms');
    }

    if (this._costUsd === undefined) {
      throw ValidationError.missingField('cost_usd');
    }

    // Build the event
    const event: TelemetryEvent = {
      event_id: this._eventId || uuidv4(),
      timestamp: this._timestamp || new Date().toISOString(),
      service_name: this._serviceName,
      trace_id: this._traceId,
      span_id: this._spanId,
      model: this._model,
      prompt: this._prompt,
      response: this._response,
      latency_ms: this._latencyMs,
      cost_usd: this._costUsd,
      metadata: this._metadata,
      errors: this._errors,
    };

    return event;
  }

  /**
   * Reset the builder to initial state
   */
  reset(): this {
    this._eventId = undefined;
    this._timestamp = undefined;
    this._serviceName = undefined;
    this._traceId = undefined;
    this._spanId = undefined;
    this._model = undefined;
    this._prompt = undefined;
    this._response = undefined;
    this._latencyMs = undefined;
    this._costUsd = undefined;
    this._metadata = {};
    this._errors = [];
    return this;
  }
}

/**
 * Helper function to create a TelemetryEvent from simple parameters
 */
export function createTelemetryEvent(params: {
  service: string;
  model: string;
  prompt: string;
  promptTokens: number;
  response: string;
  responseTokens: number;
  latencyMs: number;
  costUsd: number;
  finishReason?: string;
  traceId?: string;
  spanId?: string;
  metadata?: Record<string, string>;
  errors?: string[];
  promptEmbedding?: number[];
  responseEmbedding?: number[];
}): TelemetryEvent {
  const builder = new TelemetryEventBuilder()
    .service(params.service)
    .model(params.model)
    .prompt({
      text: params.prompt,
      tokens: params.promptTokens,
      embedding: params.promptEmbedding,
    })
    .response({
      text: params.response,
      tokens: params.responseTokens,
      finish_reason: params.finishReason || 'stop',
      embedding: params.responseEmbedding,
    })
    .latency(params.latencyMs)
    .cost(params.costUsd);

  if (params.traceId) {
    builder.traceId(params.traceId);
  }

  if (params.spanId) {
    builder.spanId(params.spanId);
  }

  if (params.metadata) {
    builder.metadata(params.metadata);
  }

  if (params.errors) {
    builder.errors(params.errors);
  }

  return builder.build();
}

File: sdks/python/sentinel_sdk/builders.py
"""Fluent builders for constructing telemetry events.

This module provides builder classes that allow for easy and readable
construction of telemetry events using method chaining.
"""

from datetime import datetime
from typing import Dict, List, Optional
from uuid import UUID, uuid4

from .events import PromptInfo, ResponseInfo, TelemetryEvent
from .exceptions import ValidationError


class TelemetryEventBuilder:
    """Fluent builder for constructing TelemetryEvent instances.

    This builder provides a convenient, readable way to construct
    telemetry events using method chaining.

    Example:
        >>> event = (TelemetryEventBuilder()
        ...     .service("my-service")
        ...     .model("gpt-4")
        ...     .prompt(text="Hello", tokens=5)
        ...     .response(text="Hi!", tokens=10, finish_reason="stop")
        ...     .latency(150.5)
        ...     .cost(0.001)
        ...     .build())
    """

    def __init__(self) -> None:
        """Initialize the builder with default values."""
        self._event_id: UUID = uuid4()
        self._timestamp: datetime = datetime.utcnow()
        self._service_name: Optional[str] = None
        self._model: Optional[str] = None
        self._trace_id: Optional[str] = None
        self._span_id: Optional[str] = None
        self._prompt: Optional[PromptInfo] = None
        self._response: Optional[ResponseInfo] = None
        self._latency_ms: Optional[float] = None
        self._cost_usd: Optional[float] = None
        self._metadata: Dict[str, str] = {}
        self._errors: List[str] = []

    def event_id(self, event_id: UUID) -> "TelemetryEventBuilder":
        """Set the event ID.

        Args:
            event_id: UUID for the event

        Returns:
            Self for method chaining
        """
        self._event_id = event_id
        return self

    def timestamp(self, timestamp: datetime) -> "TelemetryEventBuilder":
        """Set the timestamp.

        Args:
            timestamp: Event timestamp

        Returns:
            Self for method chaining
        """
        self._timestamp = timestamp
        return self

    def service(self, service_name: str) -> "TelemetryEventBuilder":
        """Set the service name.

        Args:
            service_name: Name of the service

        Returns:
            Self for method chaining
        """
        self._service_name = service_name
        return self

    def model(self, model: str) -> "TelemetryEventBuilder":
        """Set the model identifier.

        Args:
            model: Model identifier (e.g., "gpt-4", "claude-3")

        Returns:
            Self for method chaining
        """
        self._model = model
        return self

    def trace_id(self, trace_id: str) -> "TelemetryEventBuilder":
        """Set the trace ID for distributed tracing.

        Args:
            trace_id: Trace identifier

        Returns:
            Self for method chaining
        """
        self._trace_id = trace_id
        return self

    def span_id(self, span_id: str) -> "TelemetryEventBuilder":
        """Set the span ID.

        Args:
            span_id: Span identifier

        Returns:
            Self for method chaining
        """
        self._span_id = span_id
        return self

    def prompt(
        self,
        text: str,
        tokens: int,
        embedding: Optional[List[float]] = None,
    ) -> "TelemetryEventBuilder":
        """Set the prompt information.

        Args:
            text: Prompt text
            tokens: Token count
            embedding: Optional embedding vector

        Returns:
            Self for method chaining
        """
        self._prompt = PromptInfo(text=text, tokens=tokens, embedding=embedding)
        return self

    def prompt_info(self, prompt: PromptInfo) -> "TelemetryEventBuilder":
        """Set the prompt information using a PromptInfo object.

        Args:
            prompt: PromptInfo instance

        Returns:
            Self for method chaining
        """
        self._prompt = prompt
        return self

    def response(
        self,
        text: str,
        tokens: int,
        finish_reason: str,
        embedding: Optional[List[float]] = None,
    ) -> "TelemetryEventBuilder":
        """Set the response information.

        Args:
            text: Response text
            tokens: Token count
            finish_reason: Finish reason (e.g., "stop", "length")
            embedding: Optional embedding vector

        Returns:
            Self for method chaining
        """
        self._response = ResponseInfo(
            text=text,
            tokens=tokens,
            finish_reason=finish_reason,
            embedding=embedding,
        )
        return self

    def response_info(self, response: ResponseInfo) -> "TelemetryEventBuilder":
        """Set the response information using a ResponseInfo object.

        Args:
            response: ResponseInfo instance

        Returns:
            Self for method chaining
        """
        self._response = response
        return self

    def latency(self, latency_ms: float) -> "TelemetryEventBuilder":
        """Set the request latency.

        Args:
            latency_ms: Latency in milliseconds

        Returns:
            Self for method chaining
        """
        self._latency_ms = latency_ms
        return self

    def cost(self, cost_usd: float) -> "TelemetryEventBuilder":
        """Set the cost.

        Args:
            cost_usd: Cost in USD

        Returns:
            Self for method chaining
        """
        self._cost_usd = cost_usd
        return self

    def metadata(self, metadata: Dict[str, str]) -> "TelemetryEventBuilder":
        """Set metadata (replaces existing metadata).

        Args:
            metadata: Metadata dictionary

        Returns:
            Self for method chaining
        """
        self._metadata = metadata.copy()
        return self

    def add_metadata(self, key: str, value: str) -> "TelemetryEventBuilder":
        """Add a single metadata entry.

        Args:
            key: Metadata key
            value: Metadata value

        Returns:
            Self for method chaining
        """
        self._metadata[key] = value
        return self

    def errors(self, errors: List[str]) -> "TelemetryEventBuilder":
        """Set errors (replaces existing errors).

        Args:
            errors: List of error messages

        Returns:
            Self for method chaining
        """
        self._errors = errors.copy()
        return self

    def add_error(self, error: str) -> "TelemetryEventBuilder":
        """Add a single error message.

        Args:
            error: Error message

        Returns:
            Self for method chaining
        """
        self._errors.append(error)
        return self

    def build(self) -> TelemetryEvent:
        """Build the TelemetryEvent.

        Returns:
            Constructed TelemetryEvent instance

        Raises:
            ValidationError: If required fields are missing
        """
        # Validate required fields
        if self._service_name is None:
            raise ValidationError("service_name is required")
        if self._model is None:
            raise ValidationError("model is required")
        if self._prompt is None:
            raise ValidationError("prompt is required")
        if self._response is None:
            raise ValidationError("response is required")
        if self._latency_ms is None:
            raise ValidationError("latency_ms is required")
        if self._cost_usd is None:
            raise ValidationError("cost_usd is required")

        return TelemetryEvent(
            event_id=self._event_id,
            timestamp=self._timestamp,
            service_name=self._service_name,
            model=self._model,
            trace_id=self._trace_id,
            span_id=self._span_id,
            prompt=self._prompt,
            response=self._response,
            latency_ms=self._latency_ms,
            cost_usd=self._cost_usd,
            metadata=self._metadata,
            errors=self._errors,
        )


__all__ = ["TelemetryEventBuilder"]
